### PR TITLE
Add EIP-5539: Revocation List Registry

### PR DESCRIPTION
**Pull request by lleifermann**
Signed-off-by: Lauritz Leifermann <lauritz.leifermann@spherity.com>

### CODE REVIEW

File: EIPS/eip-template.md [hunk 1]
```diff
@@ -0,0 +1,235 @@
+---
+eip: <to be assigned>
```
Comment (Pandapip1) [inline EIPS/eip-template.md:2]:
```suggestion
eip: 5539
```
Don't forget to rename the file too!

File: EIPS/eip-template.md [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+title: Revocation List Registry
+description: Registry of revocation lists for revoking arbitrary data.
+author: Philipp Bolte (@strumswell), Lauritz Leifermann (@lleifermann), Dennis von der Bey (@DennisVonDerBey)
+discussions-to: <URL>
+status: Draft
+type: Standards Track
+category: ERC
+created: 2022-08-26
+requires (*optional): EIP-712
```
Comment (Pandapip1) [inline EIPS/eip-template.md:11]:
```suggestion
requires: 712
```

File: EIPS/eip-template.md [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+function batchChangeListStatuses(bool[] revokedStatuses, address namespace, bytes32 list, bytes32[] keys) public;
+```
+
+**batchChangeListStatusesSigned** ([see meta transactions](#Meta-transactions))
+**OPTIONAL** implements a function to change multiple revocation statuses in a specific namespace's revocation list with a raw signature.
+```solidity
+function batchChangeListStatusesSigned(bool[] revokedStatuses, address namespace, bytes32 list, bytes32[] keys, uint8 sigV, bytes32 sigR, bytes32 sigS) public;
+```
+
+### Owner managment
```
Comment (Pandapip1) [inline EIPS/eip-template.md:96]:
```suggestion
### Owner management
```

File: EIPS/eip-template.md [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+## Backwards Compatibility
+Not applicable
+
+## Reference Implementation
+tba
+
+## Security Considerations
+
+### Meta Transactions
+The signature of signed transactions could potentially be replayed on different chains or deployed versions of the registry implementing this ERC. This security consideration is addressed by the usage of [EIP-712](/EIPS/eip-712
```
Comment (Pandapip1) [inline EIPS/eip-template.md:229]:
```suggestion
The signature of signed transactions could potentially be replayed on different chains or deployed versions of the registry implementing this ERC. This security consideration is addressed by the usage of [EIP-712](./eip-712.md)
```

File: EIPS/eip-template.md [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+This section uses the following terms: 
+
+**`transaction signer`:** An Ethereum address that signs arbitrary data for the contract to execute **BUT** does not commit the transaction. 
+**`transaction sender`**: An Ethereum address that takes signed data from a **transaction signer** and commits it wrapped with its own signature to the smart contract.
+
+An address (**transaction signer**) **MAY** be able to deliver a signed payload off-band to another address (**transaction sender**) that initiates the Ethereum interaction with the smart contract. The signed payload **MUST** be limited to be used only once ([Signed Hash](#Signed-Hash) + [nonces](#Nonce)).
+
+#### Signed Hash
+
+The signature of the **transaction signer** **MUST** conform [EIP-712](/EIPS/eip-712), in its state that is proposed in [`ethereum/EIPs/issues/5475`](https://github.com/ethereum/EIPs/issues/5475). This helps users understand what the payload they're signing consists of & it improves the protection against replay attacks.
```
Comment (Pandapip1) [inline EIPS/eip-template.md:206]:
```suggestion
The signature of the **transaction signer** **MUST** conform [EIP-712](./eip-712.md). This helps users understand what the payload they're signing consists of & it improves the protection against replay attacks.
```
Required: External links (even to this GitHub repository) aren't allowed.
Required: Internal links must be used when applicable.

File: EIPS/eip-template.md [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+> This provides every Ethereum address a reserved space, without the need to actively claim it in the contract. Initially addresses only have owner access in their own namespace.
+
+### Why does a namespace always represent the initial owner address? 
+> The change of an owner of a list shouldn't break the link to a revocation key in it, as already existing off-chain data may depend on it. 
+
+## Backwards Compatibility
+Not applicable
+
+## Reference Implementation
+tba
```
Comment (Pandapip1) [inline EIPS/eip-template.md:224]:
```suggestion
```
Required: The Reference Implementation section is optional and should be left empty when there is no content.

File: EIPS/eip-template.md [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+
+## Security Considerations
+
```
Comment (Pandapip1) [inline EIPS/eip-template.md:227]:
```suggestion
## Security Considerations
Needs discussion.

```
Required: Security Considerations is *not* an optional section.

File: EIPS/eip-template.md (hunk 3 above)
Comment (Pandapip1) [inline EIPS/eip-template.md:96]:
```suggestion
### Owner management
```
Required: Typo

File: EIPS/eip-template.md (hunk 2 above)
Comment (Pandapip1) [inline EIPS/eip-template.md:11]:
```suggestion
requires: 712
```
Required: This is the correct format for the requires header.

File: EIPS/eip-template.md (hunk 4 above)
Comment (Pandapip1) [inline EIPS/eip-template.md:229]:
```suggestion
The signature of signed transactions could potentially be replayed on different chains or deployed versions of the registry implementing this ERC. This security consideration is addressed by the usage of [EIP-712](./eip-712.md).
```
```suggestion
The signature of the **transaction signer** **MUST** conform [EIP-712](./eip-712.md). This helps users understand what the payload they're signing consists of & it improves the protection against replay attacks.
```
Required: Internal links must be used when applicable.

File: EIPS/eip-template.md [hunk 8]
```diff
@@ -0,0 +1,235 @@
+---
+eip: <to be assigned>
+title: Revocation List Registry
+description: Registry of revocation lists for revoking arbitrary data.
+author: Philipp Bolte (@strumswell), Lauritz Leifermann (@lleifermann), Dennis von der Bey (@DennisVonDerBey)
+discussions-to: <URL>
```
Comment (Pandapip1) [inline EIPS/eip-template.md:6]: Please create a thread on https://ethereum-magicians.org/

File: EIPS/eip-template.md [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+## Rationale
+
+### Why the concept of namespaces?
+> This provides every Ethereum address a reserved space, without the need to actively claim it in the contract. Initially addresses only have owner access in their own namespace.
+
+### Why does a namespace always represent the initial owner address? 
+> The change of an owner of a list shouldn't break the link to a revocation key in it, as already existing off-chain data may depend on it. 
+
+## Backwards Compatibility
+Not applicable
```
Comment (Pandapip1) [inline EIPS/eip-template.md:221]:
```suggestion
No backward compatibility issues were found.
```
Required: Backwards Compatibility is *always* required, even if there is nothing to be compatible with.

**Review by Pandapip1 — CHANGES_REQUESTED** [review]
Here's my first pass of this.

(FYI - the reason I'm being blunt/trite here is to avoid any unnecessary back and forth. This is me apologizing in advance since it's likely that my feedback might come off as rude if you're not used to it.)

File: EIPS/eip-template.md (hunk 8 above)
Comment (lleifermann) [reply EIPS/eip-template.md:6]: Created a post there but it needs to be unblocked by a moderator first.

Comment (Pandapip1) [reply EIPS/eip-template.md:6]: Huh, that's weird and shouldn't happen. CC @jpitts

File: EIPS/eip-template.md (hunk 7 above)
Comment (Pandapip1) [reply EIPS/eip-template.md]: I missed that. My bad.

**Comment by lleifermann** [comment]
> Here's my first pass of this.
> 
> (FYI - the reason I'm being blunt/trite here is to avoid any unnecessary back and forth. This is me apologizing in advance since it's likely that my feedback might come off as rude if you're not used to it.)

Not at all! Thanks for you quick feedback on this. I've resolved most of your comments except the ethereum-magicians link (need to be allowed to post by a moderator first) and the Security Considerations Section. See my comments above.

**Comment by Pandapip1** [comment]
The security considerations bit is my fault. Completely missed all the text under it.

File: EIPS/eip-template.md (hunk 8 above)
Comment (lleifermann) [reply EIPS/eip-template.md]: To be fair i never had an account there so i may have triggered some spam system by immediately posting 😄. Here's the account -> https://ethereum-magicians.org/u/lleifermann/

File: EIPS/eip-5539.md [hunk 10]
```diff
@@ -0,0 +1,232 @@
+---
+eip: 5539
+title: Revocation List Registry
+description: Registry of revocation lists for revoking arbitrary data.
+author: Philipp Bolte (@strumswell), Lauritz Leifermann (@lleifermann), Dennis von der Bey (@DennisVonDerBey)
+discussions-to: <URL>
```
Comment (Pandapip1) [inline EIPS/eip-5539.md:6]: Required: Please create a thread on Ethereum magicians

Comment (lleifermann) [reply EIPS/eip-5539.md:6]: As far as I can tell my account is still blocked in the forum.

Comment (Pandapip1) [reply EIPS/eip-5539.md:6]: Strange... could you send a screenshot?

**Comment by DennisVonDerBey** [comment]
Added the discussion link now with a topic created on my account.

File: EIPS/eip-5539.md (hunk 10 above)
Comment (DennisVonDerBey) [reply EIPS/eip-5539.md]: it's fixed now

**Review by Pandapip1 — APPROVED** [review]

**Review by Pandapip1 — CHANGES_REQUESTED** [review]
Please fix all the internal links

File: EIPS/eip-5539.md [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+The signature of the **transaction signer** **MUST** conform [EIP-712](./eip-712.md). This helps users understand what the payload they're signing consists of & it improves the protection against replay attacks.
+
+#### Nonce
+
+This EIP **RECOMMENDS** the use of a **dedicated nonce mapping** for meta transactions. If the signature of the **transaction sender** and its meta contents are verified, the contract increases a nonce for this **transaction signer**. This effectively removes the possibility for any other sender to execute the same transaction again with another wallet. 
+
+## Rationale
+
+### Why the concept of namespaces?
+> This provides every Ethereum address a reserved space, without the need to actively claim it in the contract. Initially addresses only have owner access in their own namespace.
```
Comment (xinbenlv) [inline EIPS/eip-5539.md:215]:
```suggestion
This provides every Ethereum address a reserved space, without the need to actively claim it in the contract. Initially addresses only have owner access in their own namespace.
```
Suggest removing the `>` unless there is any particular reason.

File: EIPS/eip-5539.md [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+discussions-to: https://ethereum-magicians.org/t/eip-5539-revocation-list-registry/10573
+status: Draft
+type: Standards Track
+category: ERC
+created: 2022-08-26
+requires: 712
+---
+
+## Abstract
+This EIP proposes a set of methods and standards for an RBAC-enabled registry of indicators aimed for usage in revocations.
```
Comment (xinbenlv) [inline EIPS/eip-5539.md:15]: [Editorial] consider spell out `RBAC` at its first occurance

Comment (Pandapip1) [reply EIPS/eip-5539.md:15]: I strongly agree here. Acronyms and jargon should be defined unless well-known.

Comment (strumswell) [reply EIPS/eip-5539.md]: I simplified it.

File: EIPS/eip-5539.md [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+
+This EIP **RECOMMENDS** the use of a **dedicated nonce mapping** for meta transactions. If the signature of the **transaction sender** and its meta contents are verified, the contract increases a nonce for this **transaction signer**. This effectively removes the possibility for any other sender to execute the same transaction again with another wallet. 
+
+## Rationale
+
+### Why the concept of namespaces?
+> This provides every Ethereum address a reserved space, without the need to actively claim it in the contract. Initially addresses only have owner access in their own namespace.
+
+### Why does a namespace always represent the initial owner address? 
+> The change of an owner of a list shouldn't break the link to a revocation key in it, as already existing off-chain data may depend on it.
```
[the remark this answers is not in the document]
Comment (strumswell) [reply EIPS/eip-5539.md]: Done 👍🏻

**Comment by strumswell** [comment]
> Please fix all the internal links

They should be fixed now!

**Comment by strumswell** [comment]
@Pandapip1 can you re-review so we can get this merged?

**Review by Pandapip1 — APPROVED** [review]

**Pull request merged**